### PR TITLE
feat(rds): allow excludedCharacters for addRotation*

### LINKED ISSUE #4144: Specify properties of the Secret as part of Database instance - Description, ExcludeCharacters
When doing `new rds.DatabaseInstance(this, 'AppDb', { ... })`, a secret is automatically generated with the database's connection info. This is super useful, but not quite sufficient for my use case. I'd like to be able to specify the secret properties when creating the RDS instance.

### Use Case

I need to specify some other characters to exclude from the generated password (specifically `;`, since I'll be assembling a semicolon-delimited connection string like `Host=abc.com;Port=5432;Username=admin;Password=p@ssw0rd;`).

### Proposed Solution

Add a `secretProperties` property to the RDS `DatabaseInstanceProps` that mixes in any specified options with the defaults when creating the DB secret, e.g.:

```ts
const appDbInstance = new DatabaseInstance(this, 'AppDb', {
	databaseName: 'AppDb',
	engine: DatabaseInstanceEngine.POSTGRES,
	engineVersion: '10.9',
	secretProperties: {
		excludeCharacters: '"@/\\;'
	}
});
```

### Other

I would override this manually, but trying this only adds a property to the `SecretTargetAttachment` not the Secret itself:

```ts
const cfnAppSecret = appDbInstance.secret!.node.defaultChild as CfnSecretTargetAttachment;

cfnAppSecret.addOverride('Properties.GenerateSecretString.ExcludeCharacters', '"@/\\;');
```

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by ahammond**
fix #4144

Configurable `excludedCharacters` for addRotation*

This adds a sane default for addRotationSingleUser.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 1]
```diff
@@ -560,8 +560,11 @@ export class DatabaseCluster extends DatabaseClusterNew {
    *
    * @param [automaticallyAfter=Duration.days(30)] Specifies the number of days after the previous rotation
    * before Secrets Manager triggers the next automatic rotation.
+   *
+   * @param [excludeCharacters=' ;+%{}`/"\\#\'@'] Specifies characters to not include in generated passwords.
+   * Defaults to the superset of characters which will break DMS endpoints and characters which cause problems in shell scripts.
    */
-  public addRotationSingleUser(automaticallyAfter?: Duration): secretsmanager.SecretRotation {
+  public addRotationSingleUser(automaticallyAfter?: Duration, excludedCharacters?: string): secretsmanager.SecretRotation {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:567]: Can we actually add an interface here? This code:

```ts
instance.addRotationSingleUser(
  cdk.Duration.days(7), // automaticallyAfter defaults to 30 days
  'abc', // excludedCharacters defaults to ' ;+%{}`/"\\#\'@', the combination of characters that break DMS and that cause problems in shell scripts.
);
```

looks extremely cryptic to me; I'd rather see this:

```ts
instance.addRotationSingleUser(cdk.Duration.days(7), { // automaticallyAfter defaults to 30 days
  excludedCharacters: 'abc', // excludedCharacters defaults to ' ;+%{}`/"\\#\'@', the combination of characters that break DMS and that cause problems in shell scripts.
});
```

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 2]
```diff
@@ -24,6 +24,13 @@ export interface DatabaseSecretProps {
    * @default - no master secret information will be included
    */
   readonly masterSecret?: secretsmanager.ISecret;
+
+  /**
+   * Characters to not include in the generated password.
+   *
+   * @default '"@/\\'
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/database-secret.ts:31]: Hmm. Why is the default here different than the default for rotation?

File: packages/@aws-cdk/aws-rds/lib/props.ts [hunk 3]
```diff
@@ -147,6 +147,13 @@ export interface RotationMultiUserOptions {
    * @default Duration.days(30)
    */
   readonly automaticallyAfter?: Duration;
+
+  /**
+   * Specifies characters to not include in generated passwords.
+   *
+   * @default -
+   */
+  readonly excludedCharacters?: string;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/props.ts:156]: Is this used anywhere?

File: packages/@aws-cdk/aws-rds/test/integ.cluster-rotation.lit.expected.json [hunk 4]
```diff
@@ -807,7 +807,8 @@
               "DatabaseRotationSingleUserSecurityGroupAC6E0E73",
               "GroupId"
             ]
-          }
+          },
+          "excludeCharacters": " ;+%{}@'\"`/\\#"
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/integ.cluster-rotation.lit.expected.json:811]: A unit test would be nice as well 🙂

**Review by skinny85 — DISMISSED** [review]
Thanks for the contribution @ahammond ! A few comments/questions to start 🙂.

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 2 above)
Comment (ahammond) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:31]: Legacy. Should I standardize? Also, I was wondering if I should create some static strings under aws-dms or somewhere to track these.

Comment (ahammond) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:31]: Also... should I simply pass null if given null and then use the default in secretsmanager?

Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:31]: What's the default in SecretsManager?

File: packages/@aws-cdk/aws-rds/lib/props.ts (hunk 3 above)
Comment (ahammond) [reply packages/@aws-cdk/aws-rds/lib/props.ts]: It is passed with the rest of the options through to the `secretsmanager.SecretRotation()` call at https://github.com/ahammond/aws-cdk/blob/excluded-characters-in-aurora-passwords/packages/%40aws-cdk/aws-rds/lib/cluster.ts#L593

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 1 above)
Comment (ahammond) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts]: Yeah, I didn't like it, but I didn't want to break anything. Adding options and not including the duration felt ugly, so...

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 2 above)
Comment (ahammond) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts]: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-secretsmanager/lib/secret-rotation.ts#L200

No excluded characters.

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 5]
```diff
@@ -43,7 +50,9 @@ export class DatabaseSecret extends secretsmanager.Secret {
           masterarn: props.masterSecret?.secretArn,
         }),
         generateStringKey: 'password',
-        excludeCharacters: '"@/\\',
+        excludeCharacters: props.excludeCharacters === undefined
+          ? secretsmanager.ProblemCharacters.AWS_DMS + secretsmanager.ProblemCharacters.SHELL
+          : props.excludeCharacters,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/database-secret.ts:55]:
```suggestion
        excludeCharacters: props.excludeCharacters ?? (secretsmanager.ProblemCharacters.AWS_DMS + secretsmanager.ProblemCharacters.SHELL),
```

File: packages/@aws-cdk/aws-rds/lib/props.ts [hunk 6]
```diff
@@ -120,6 +120,23 @@ export interface Login {
...
 }
 
+/**
+ * Options to add the multi user rotation
+ */
+export interface RotationSingleUserOptions {
+  /**
+   * Specifies the number of days after the previous rotation
+   * before Secrets Manager triggers the next automatic rotation.
+   * @default Duration.days(30)
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/props.ts:130]: Missing empty line before this line.

File: packages/@aws-cdk/aws-rds/lib/props.ts [hunk 7]
```diff
@@ -120,6 +120,23 @@ export interface Login {
...
+   */
+  readonly automaticallyAfter?: Duration;
+  /**
+  * Specifies characters to not include in generated passwords.
+  * @default secretsmanager.ProblemCharacters.AWS_DMS+secretsmanager.ProblemCharacters.SHELL
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/props.ts:135]: Missing empty line before this line.

File: packages/@aws-cdk/aws-secretsmanager/test/test.problem-secrets.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { Test } from 'nodeunit';
+import * as secretsmanager from '../lib';
+
+export = {
+  'ProblemCharacters exist'(test: Test) {
+    expect(secretsmanager.ProblemCharacters.AWS_DMS).toBeDefined;
+    expect(secretsmanager.ProblemCharacters.SHELL).toBeDefined;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/test/test.problem-secrets.ts:7]: I would remove this test, it doesn't add any value.

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts [hunk 9]
```diff
@@ -5,6 +5,14 @@ import { ResourcePolicy } from './policy';
 import { RotationSchedule, RotationScheduleOptions } from './rotation-schedule';
 import * as secretsmanager from './secretsmanager.generated';
 
+/**
+ * Character Sets which cause problems for other services when used in passwords.
+ */
+export enum ProblemCharacters {
+  AWS_DMS = ' %+:;{}',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:12]: A few things.

1. You're often concatenating these. Would these be better served as `string` constants?

```ts
export class ProblemCharacters {
  public static readonly AWS_DMS = ' %+:;{}';

  public static readonly SHELL = '~`#$&*()|[]{};<>?!\'\\';
}
```

2. Some docs on these would be nice.
3. I see some characters repeat. Does `SecretManager` handle that correctly? Should we de-duplicate them in the construct before generating a CFN template with duplicates?

**Review by skinny85 — CHANGES_REQUESTED** [review]
Thanks @ahammond for the contribution! I like the" breaking change" here. Makes the interface more consistent between `addRotationSingleUser` and `addRotationMultiUser`.

Should the changes be also added to `Instance`, and its `addRotationSingleUser` method?

Also, make sure to `yarn build` the SecretsManger locally, as it doesn't build currently.

**Comment by skinny85** [comment]
@ahammond I've taken this over, in https://github.com/aws/aws-cdk/pull/10583 . Would appreciate your input whether that PR correctly fulfills the intent of the original issue!

**Comment by ahammond** [comment]
@skinny85 That's a good solution. I still think there might be value in some kind of way to keep track of characters that break different kinds of passwords (dms, bash, etc), but... that's a problem for another PR.

**Pull request closed without merging**